pageserver/src/walingest.rs

```diff
@@ -636,16 +634,17 @@ impl<'a> WalIngest<'a> {
                 forknum: VISIBILITYMAP_FORKNUM,
             };
 
-            // FIXME: Like with the FSM above, the logic to truncate the VM
-            // correctly has not been implemented. Just zap it away completely,
-            // always. Unlike the FSM, the VM must never have bits incorrectly
-            // set. From a correctness point of view, it's always OK to clear
-            // bits or remove it altogether, though.
-            if rec.blkno != 0 {
-                info!("Partial truncation of VM is not supported");
+            let mut vm_page_no = rec.blkno / pg_constants::VM_HEAPBLOCKS_PER_PAGE;
+            if rec.blkno % pg_constants::VM_HEAPBLOCKS_PER_PAGE != 0 {
+                // Tail of last remaining vm page has to be zeroed.
+                // We are not doing it here: there should be FPI wal-record for this page stored sometimes laters
+                vm_page_no += 1;
+            }
```

Comment:
This seems unsafe. Firstly, there's no guarantee that the FPI record follows. PostgreSQL might've crashed before the record was written out. Secondly, you might be doing point-in-time query between the two records.

Reply:
Content of FSM/VM pages is non critical. It is expected to be lost.
By default changes of FSM/VM are no wal-logged.
So it seems to be normal in Vanilla that after abnormal termination we can have partly updated map.
I do no think here some difference in Neon and Vanilla behavior. Yes, in Neon we are separately updating relation size. But what wrong can happen if length o relation is updated and content - not?

Reply:
It's safe to clear bits in the VM, but if a bit is set in the VM, it's important that all the tuples on the pages are visible, or you can get incorrect results in queries. We mustn't leave behind bits that are set when they shouldn't be.

Reply:
I have to agree with you - it can cause problems.
Actually it is harmless to make some non-zeroed bit in VM map beyond end of the relation.
As far as there are no records o this pages, marking them as all-visible in visibility map should not be a problem.
Moreover - it is even correct: all records on empty page are visible:)

The problem arises when we try to isert some record o this page. Insert WAL record has flag `XLH_INSERT_ALL_VISIBLE_CLEARED`, but it is set only if page was mrked as all-visible.
And new page definitely has no such mark.

But I think that it is not so difficult to fix the problem.
Truncation is performed by `RelationTruncate` which in turn calls `visibilitymap_prepare_truncate`. 
And it contains the following code:
```
		/*
		 * Truncation of a relation is WAL-logged at a higher-level, and we
		 * will be called at WAL replay. But if checksums are enabled, we need
		 * to still write a WAL record to protect against a torn page, if the
		 * page is flushed to disk before the truncation WAL record. We cannot
		 * use MarkBufferDirtyHint here, because that will not dirty the page
		 * during recovery.
		 */
		MarkBufferDirty(mapBuffer);
		if (!InRecovery && RelationNeedsWAL(rel) && XLogHintBitIsNeeded())
			log_newpage_buffer(mapBuffer, false);
```
So if we have `wal_log_hints` set, then we will insert in WAL FPI record with truncated page image  before `XLOG_SMGR_TRUNCATE `. No inconsistency is possible in this case.
But we do not want to switch on `wal_log_hints`. what can we do is just remove `XLogHintBitIsNeeded()` preicate here  or do something like this:
```
	if (!InRecovery && RelationNeedsWAL(rel) && (XLogHintBitIsNeeded() || IsNeonMode)
```

where is `IsNeonMode` (suggest better name) is flag set in neon exntesion.
I am not sure that we need such flag in our Postgres fork. Because I do not see any reasons to use our postgres in "non-Neon" mode. 

Alternative solution is to implement correct zeroing of the tail of last bitmap page in walingest.rs.
Bit if it possible to avoid code duplication (convert C code to Rust), I prefer not to do it.

What do you think?

Reply:
> As far as there are no records o this pages, marking them as all-visible in visibility map should not be a problem.
> Moreover - it is even correct: all records on empty page are visible:)

:-). Not quite though: the actual rule is that the if a heap page has `!PD_ALL_VISIBLE`, then the visibility map bit must not be set. A newly-initialized page does not have its `PD_ALL_VISIBLE` flag set.

Reply:
> But I think that it is not so difficult to fix the problem. Truncation is performed by `RelationTruncate` which in turn calls `visibilitymap_prepare_truncate`. And it contains the following code:
> 
> ```
> 		/*
> 		 * Truncation of a relation is WAL-logged at a higher-level, and we
> 		 * will be called at WAL replay. But if checksums are enabled, we need
> 		 * to still write a WAL record to protect against a torn page, if the
> 		 * page is flushed to disk before the truncation WAL record. We cannot
> 		 * use MarkBufferDirtyHint here, because that will not dirty the page
> 		 * during recovery.
> 		 */
> 		MarkBufferDirty(mapBuffer);
> 		if (!InRecovery && RelationNeedsWAL(rel) && XLogHintBitIsNeeded())
> 			log_newpage_buffer(mapBuffer, false);
> ```
> 
> So if we have `wal_log_hints` set, then we will insert in WAL FPI record with truncated page image before `XLOG_SMGR_TRUNCATE `. No inconsistency is possible in this case. But we do not want to switch on `wal_log_hints`. what can we do is just remove `XLogHintBitIsNeeded()` preicate here or do something like this:
> 
> ```
> 	if (!InRecovery && RelationNeedsWAL(rel) && (XLogHintBitIsNeeded() || IsNeonMode)
> ```
> 
> where is `IsNeonMode` (suggest better name) is flag set in neon exntesion. I am not sure that we need such flag in our Postgres fork. Because I do not see any reasons to use our postgres in "non-Neon" mode.

Hmm, that works for WAL generated with patched postgres. It will still not do the right thing when digesting WAL from vanilla PostgreSQL, however. It would be good to maintain that compatibility.

> Alternative solution is to implement correct zeroing of the tail of last bitmap page in walingest.rs. Bit if it possible to avoid code duplication (convert C code to Rust), I prefer not to do it.

Yeah, it's a fair amount of hairy code :-(.

Third option would be to overwrite the whole last bitmap page with zeros in walingest.rs. That should be safe. Although that's what we thought about just truncating the VM too, and that turned out to be a bad idea, so I feel a little scared about that.

It would be nice to change this in PostgreSQL. It's weird that we WAL-log the full-page image separately like that. It would seem more natural to include that full-page image in the XLOG_SMGR_TRUNCATE record itself. And when we don't log a full page image, it would seem natural to include the byte and bit ranges that are being zeroed in the XLOG_SMGR_TRUNCATE record, instead of re-doing the math in the redo function.

Reply:
> It will still not do the right thing when digesting WAL from vanilla PostgreSQL,
 
But our wal-redo in any case will not be able to reply WAL generated by Vanilla because of format difference (cid,...). So I do not think that we loose something by making this change.

> Third option would be to overwrite the whole last bitmap page with zeros in walingest.rs.

Yes, it is possible and will be not difficult. But "aesthetically" I do not like this option: it looks too compromising. I prefer that we either do nothing with bitmap content, either correctly clear it. But here we update bitmap, but not precisely... But if you think that it is the nest solution, then I will implement it: just adding few lines of code.